### PR TITLE
User input subhrticks

### PR DESCRIPTION
**Pull request by galanCA**
## Description

Expose the nSubhrTicks for the user to control HPWHSim timestep.

## Author Progress Checklist:

- [x] Open draft pull request
    - [x] Make title clearly understandable in a standalone change log context
    - [x] Assign yourself the issue
    - [x] Add at least one label (enhancement, bug, or maintenance)
    - [ ] Link the issue(s) addressed by this PR (under "Development" in the sidebar menu) 
- [x] Make code changes (if you haven't already)
- [ ] Self-review of code
    - [x] My code follows the style guidelines of this project
    - [ ] I have added comments to my code, particularly in hard-to-understand areas
    - [ ] I have only committed the necessary changes for this fix or feature
    - [ ] I have made corresponding changes to the documentation
    - [ ] My changes generate no new warnings
    - [ ] I have ensured that my fix is effective or that my feature works as intended by:
        - [x] exercising the code changes in the test framework, and/or
        - [ ] manually verifying the changes (as explained in the the pull request description above)
    - [ ] My changes pass all local tests
    - [ ] My changes successfully passes CI checks
    - [ ] Add any unit test for proof and documentation.
    - [ ] Merge in main branch and address resulting conflicts and/or test failures.
- [ ] Move pull request out of draft mode and assign reviewers
- [ ] Iterate with reviewers until all changes are approved
    - [ ] Make changes in response to reviewer comments
    - [ ] Merge in main branch and address resulting conflicts and/or test failures.
    - [ ] Re-request review in GitHub

## Reviewer Checklist:

 - [x] Read the pull request description
 - [x] Perform a code review on GitHub
 - [x] Confirm all CI checks pass and there are no build warnings
 - [ ] Pull, build, and run automated tests locally
 - [ ] Perform manual tests of the fix or feature locally
 - [x] Add any review comments, if applicable
 - [ ] Submit review in GitHub as either
     - [ ] Request changes, or
     - [x] Approve
 - [x] Iterate with author until all changes are approved

### CODE REVIEW

File: src/CNCULT2.CPP [hunk 1]
```diff
@@ -377,15 +377,16 @@ RC TOPRAT::tp_SetCheckTimeSteps()	// initialize timesteps
 	tp_subhrDur = 1.f / tp_nSubSteps;			// subhr duration
 
 	// substep ticks = even shorter steps for e.g. HPWH simulation
-	tp_nSubhrTicks = max(1, 60 / tp_nSubSteps);
+	if (tp_nSubhrTicks == 0) // Not set
```
Comment (nealkruis) [inline src/CNCULT2.CPP:380]: See the `IsSet` method.

File: doc/src/records/top-members.md [hunk 2]
```diff
@@ -81,6 +81,19 @@ Number of subhour steps used per hour in the simulation. 4 is the time-honored v
   required: "No",
   variability: "constant") %>
 
+**nSubhrTicks=*int***
+
+Number of subhour ticks used per *nSubSteps* in the simulation.
+
+*Warning:* nSubhrTicks is in a experimental phase.
```
Comment (tanaya-bigladder) [inline doc/src/records/top-members.md:88]: "an" experimental phase

File: src/CNCULT2.CPP [hunk 3]
```diff
@@ -377,15 +377,20 @@ RC TOPRAT::tp_SetCheckTimeSteps()	// initialize timesteps
...
 	tp_tickDurHr = 1. / (tp_nSubhrTicks * tp_nSubSteps);
 	tp_tickDurMin = tp_tickDurHr * 60.;
-
+	/*
 	if (tp_tickDurMin != 1. && bUsesTicks)
 		rc |= oer("nSubSteps=%d is invalid because it yields a non-integral substep duration (%0.3f min)."
 			   "\n    An integral duration is required when DHWSYS(s) are present.",
 			tp_nSubSteps, tp_tickDurMin);
-
+	*/
```
Comment (tanaya-bigladder) [inline src/CNCULT2.CPP:393]: We should deprecate the use of C-style comments, preferring "`//`". Although, better than both for this kind of code chunk is to preprocess it out with `#if 0`. (Other places in the code base that use this method, also add a comment after the `#if 0` that gives the date and reason for deprecating the code.)

**Review by tanaya-bigladder — CHANGES_REQUESTED** [review]

**Review by tanaya-bigladder — APPROVED** [review]

File: doc/src/records/top-members.md (hunk 2 above)
Comment (nealkruis) [reply doc/src/records/top-members.md]: "phase" doesn't really feel like the right term. Instead:

Note: This input is currently used only for experimental purposes.

Comment (galanCA) [reply doc/src/records/top-members.md]: I made those changes as well as changed the warning description when using nSubHrTicks.

File: src/CNCULT2.CPP [hunk 4]
```diff
@@ -377,15 +377,21 @@ RC TOPRAT::tp_SetCheckTimeSteps()	// initialize timesteps
 	tp_subhrDur = 1.f / tp_nSubSteps;			// subhr duration
 
 	// substep ticks = even shorter steps for e.g. HPWH simulation
-	tp_nSubhrTicks = max(1, 60 / tp_nSubSteps);
+	if (!IsSet(tp_nSubhrTicks)) { // Not set
+		tp_nSubhrTicks = max(1, 60 / tp_nSubSteps); // Default
+	}
+	else if (tp_nSubhrTicks != max(1, 60 / tp_nSubSteps)) {
+		warn("This input is currenlty used only for experimental purposes");
```
Comment (nealkruis) [inline src/CNCULT2.CPP:384]: @galanCA have you triggered this warning? How does it appear to a user? Is there enough context to understand the message?

**Review by nealkruis — CHANGES_REQUESTED** [review]

**Comment by galanCA** [comment]
I opted to use orWarn as a warning function because I thought it gave more information. oWarn gave the name of the CSE file the record and the warning. orWarn explain when during runtime the warning happened the record and the warning.

**Comment by galanCA** [comment]
This is the output shown for oWarn

oWarn

This is the output shown for orWarn

orWarn

**Comment by nealkruis** [comment]
But is this an input warning or a runtime warning (i.e., something that you encounter during a simulation after the input has been processed)?

**Comment by galanCA** [comment]
An empty destructor is needed for DHWTICK as it is mentioned in this [post](https://www.quora.com/Is-it-recommended-to-create-an-empty-destructor-of-a-class). This is so the parent class knows that there is a destructor in the child class.

**Review by nealkruis — APPROVED** [review]
This is good enough how it is. We'll resolve the potential HPWHSim issues separately.

**Pull request merged**